heroku_connect/models.py

```diff
@@ -0,0 +1,274 @@
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#understanding-the-trigger-log
+    .. _error handling:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#write-errors
+
+    """
+
+    # I18N / TRANSLATIONS:
+    #
+    # Field names and choices don't use translations, because they refer to (English)technical
+    # terms in Heroku Connect's manual. That connection should be preserved. Trigger Log Models
+    # are for technical folks and not intended to be user-facing.
+
+    # read-only fields
+    id = models.BigIntegerField(primary_key=True, editable=False)
+    created_at = models.DateTimeField(editable=False, null=True)
+    updated_at = models.DateTimeField(editable=False, null=True)
+    processed_at = models.DateTimeField(editable=False, null=True)
+    table_name = models.CharField(max_length=128, editable=False)
+    record_id = models.BigIntegerField(editable=False)
+    sf_id = models.CharField(max_length=18, editable=False, null=True, db_column='sfid')
+    action = models.CharField(max_length=7, editable=False, choices=TRIGGER_LOG_ACTION_CHOICES)
+    sf_message = models.TextField(editable=False, null=True, blank=True)
+
+    # TODO: these are more useful as HStoreFields (if 'django.contrib.postgres' in INSTALLED_APPS)
+    # leave them as textfields for now
+    _values = models.TextField(editable=False, null=True, blank=True, db_column='values')
+    _old = models.TextField(editable=False, null=True, blank=True, db_column='old')
+
+    # editable fields
+    state = models.CharField(max_length=8, null=False, blank=False,
+                             choices=TRIGGER_LOG_STATE_CHOICES)
+
+    objects = models.Manager.from_queryset(TriggerLogQuerySet)()
+
+    class Meta:
+        abstract = True
+        managed = False
+        get_latest_by = 'created_at'
+        ordering = ('id',)
+
+    is_archived = False
+
+    @classmethod
+    def capture_insert_from_model(cls, table_name, record_id, *, exclude_fields=()):
+        """
+        Create a fresh insert record from the current model state in the database.
+
+        For read-write connected models, this will lead to the attempted creation of a
+        corresponding object in Salesforce.
+
+        Args:
+            table_name: The name of the table backing the connected model (without schema)
+            record_id: The primary id of the connected model
+            exclude_fields: The names of fields that will not be included in the write record
+
+        Returns:
+            A list of the created TriggerLog entries (usually one).
+
+        Raises:
+            LookupError: if ``table_name`` does not belong to a connected model
+
+        """
+        exclude_cols = ()
+        if exclude_fields:
+            model_cls = get_connected_model_for_table_name(table_name)
+            exclude_cols = cls._fieldnames_to_colnames(model_cls, exclude_fields)
+
+        raw_query = sql.SQL("""
```

Comment:
👏

Reply:
Please show this to the others on Friday.

Reply:
You mean, how to be safe with `sql.SQL`? Sure!

Reply:
yes